go-controller/pkg/node/iptables/iptables.go

```diff
@@ -17,12 +17,79 @@ type Rule struct {
 	Protocol iptables.Protocol
 }
 
+// AddRulesFiltered adds the given rules to iptables.
+// filter is a map[table][chain] of valid tables/chains to use for filtering rules to be added.
+func AddRulesFiltered(rules []Rule, append bool, filter map[string]map[string]bool) error {
+	addErrors := errors.New("")
+	var err error
+	var ipt util.IPTablesHelper
+	var exists bool
+
+	// stores valid table chains and whether they were already created or not
+	// key is ip protocol, table, chain
+	createdChains := map[iptables.Protocol]map[string]map[string]bool{
+		iptables.ProtocolIPv4: make(map[string]map[string]bool),
+		iptables.ProtocolIPv6: make(map[string]map[string]bool),
+	}
+
+	for _, r := range rules {
+		if _, ok := filter[r.Table][r.Chain]; !ok {
+			klog.V(5).Infof("Ignoring processing rule in table: %s, chain: %s with args: \"%s\" for protocol: %v ",
+				r.Table, r.Chain, strings.Join(r.Args, " "), r.Protocol)
+			continue
+		}
+		klog.V(5).Infof("Adding rule in table: %s, chain: %s with args: \"%s\" for protocol: %v ",
+			r.Table, r.Chain, strings.Join(r.Args, " "), r.Protocol)
+		if ipt, err = util.GetIPTablesHelper(r.Protocol); err != nil {
+			addErrors = errors.Wrapf(addErrors,
+				"Failed to add iptables %s/%s rule %q: %v", r.Table, r.Chain, strings.Join(r.Args, " "), err)
+			continue
+		}
+		if _, ok := createdChains[r.Protocol][r.Table][r.Chain]; !ok {
+			klog.Infof("Creating table: %s chain: %s", r.Table, r.Chain)
+			if err = ipt.NewChain(r.Table, r.Chain); err != nil {
+				klog.V(5).Infof("Chain: \"%s\" in table: \"%s\" already exists, skipping creation: %v",
+					r.Chain, r.Table, err)
+			}
+			// we assume an error means it was already created
```

Comment:
shouldn't you be checking for the error here before you insert into the map?
semantically speaking https://github.com/coreos/go-iptables/blob/65c67c9f46e6fe231f62b9794a300c3c3dd1cc80/iptables/iptables.go#L481 says `If the chain already exists, it will result in an error.` but what if chain did not exist and NewChain tried to add it and it failed while adding it? how about that error? is that possible?

Reply:
sure its possible. This code already existed and I was just lazy and didn't fix it. I'll add another commit and fix it.

Reply:
I'm actually going to fix NewChain usage in another PR. It's used all over the place in our code. It really should have been handled in the coreos iptables library. I don't think it needs to be part of this PR because that was pre-existing code and its really fixing a different issue. Hope that is OK.